<h2>Predicting Titanic Survivors with Machine Learning</h2>

<h3>Objetive</h3>

The goal of this project is to develop a machine learning model to predict whether a passenger on the Titanic survived or not, based on various features such as age, gender, class, and ticket fare.

<h3>Dataset</h3>

The dataset used is the famous Titanic dataset, containing information about passengers, including whether they survived or not. It includes features such as age, sex, ticket class, and embarkation point.
The dataset can be downloaded here: https://www.kaggle.com/competitions/titanic/data

#References

https://www.kaggle.com/code/startupsci/titanic-data-science-solutions/notebook
https://www.kaggle.com/code/louyuechen0122/titanic-analysis
https://www.kaggle.com/code/eliassaker/titanic-survival-chances

<h3>About the data</h3>

<p><strong>According to the dataset owner (Kaggle):</strong></p>

<p>The <b>training set</b> should be used to build your machine learning models. For the training set, we provide the outcome (also known as the “ground truth”) for each passenger. Your model will be based on “features” like passengers’ gender and class. You can also use feature engineering to create new features.</p>

<p>The <b>test set</b> should be used to see how well your model performs on unseen data. For the test set, we do not provide the ground truth for each passenger. It is your job to predict these outcomes. For each passenger in the test set, use the model you trained to predict whether or not they survived the sinking of the Titanic.</p>

<h2>Step 1 - Problem Identification</h2>

<p><b>Some Relevant information about the disaster:</b></p>

<ul>
    <li>On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. Translated 32% survival rate.</li>
    <li>One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew.</li>
    <li>Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.</li>
</ul>

<p><b>Problem to be solved:</b> "Knowing from a training set of samples listing passengers who survived or did not survive the Titanic disaster, can our model determine based on a given test dataset not containing the survival information, if these passengers in the test dataset survived or not."</p>

<h2>Step 2 - Data Collection</h2>

<p>Before acquiring the dat, we will import the needed libraries to proceed with the project</p>

In [47]:
# data mining and analysis
import pandas as pd #import pandas to deal with dataframes
import numpy as np #import numpy to work with math functions

# statistics
import scipy.stats

# visualition of data
import matplotlib.pyplot as plt # library that allows you to build graphs

# machine learning algorithms (from sklearn library)
from sklearn.tree import DecisionTreeClassifier # algorithm based on decision tree
from sklearn.linear_model import LinearRegression # algorithm based on linear regression
from sklearn.linear_model import LogisticRegression # algorithm based on logistics regression
from sklearn.neighbors import KNeighborsClassifier # algorithms based on k-nearest neighbors
from sklearn.svm import SVC, LinearSVC # algorithm based on Support Vector Classification
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron # neural network with 1 neuron
from sklearn.linear_model import SGDClassifier 

<p>With the libraries imported, it is possible to proceed with the data acquisition</p>

In [36]:
# using the read_csv from Pandas to acquire the data
train_df = pd.read_csv('./datasets/train.csv')
test_df = pd.read_csv('./datasets/test.csv')

In [5]:
# Show the first 5 rows of train_df
train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
# Show the first 5 rows of test_df
test_df.head(5)

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

<h2>Step 3 - Data Exploration and Preparation</h2>

<p>To understand how to prepare the dataset, it is necessary to analyze by describing data </p>

<p>Lets see the columns names: </p>

In [7]:
display(train_df.columns.values)

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

<p>By looking the features (columns names), it is possible to separete the categorical and numerical data (continuous or discrete).</p>

<ul>
    <li><b>Categorical:</b> Survived, Sex, Embarked and Pclass (Pclass is also an ordinal data)</li>
    <li><b>Continuous: </b>Age, Fare. Discrete: SibSp, Parch</li>
    <li><b>DIscrete: </b> SibSp, Parch</li>
</ul>

<p>Now, lets see the data type of each feature:</p>

In [12]:
display(train_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


None

<p>We can see that "Age", "Cabin" and "Embarked" contains null values</p>

<p>Lets use Describe to see more info:</p>

In [19]:
display(train_df.describe())

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [20]:
train_df.describe(include=['O'])

Name   Sex Ticket        Cabin Embarked
count                              891   891    891          204      889
unique                             891     2    681          147        3
top     Gustafsson, Mr. Anders Vilhelm  male   1601  C23 C25 C27        S
freq                                 1   577      7            4      644

<h3>Assumptions based on first exploration of the data:</h3>

<p>1 - We need to complete the "Age" data, as it is probably correleted to survival</p>
<p>2 - There are a lot of duplicate values for Ticket (891 passengers and 681 unique values for Tickets (23% of duplicates values)), however Ticker may not be correleted to survival, so we can drop this feature </p>
<p>3 - Does not make sense to use "Name" to predict survival, so we can drop the feature</p>
<p>4 - Does not make sense to use "PassengerID" to predict survival, so we can drop the feature</p>
<p>5 - "Cabin" could be a relevant feature, however there are only 23% of available information (204 of 891), so seems to be better drop this feature</p>
<p>6 - Maybe it is a good approach create a new feature named family containing thee sum of "SibSp" and "Parch", this way we will have a feature with the total of family member on board the ship</p>

<h3>Preparing the data:</h3>

<p>1 - We need to complete the "Age" data, as it is probably correleted to survival</p>

In [37]:
# I will complete this missing value of the mean of Age feature (looks like a good approach for age)
train_df['Age'] = train_df['Age'].fillna(train_df['Age'].mean()) #filling the age of train dataset
test_df['Age'] = test_df['Age'].fillna(test_df['Age'].mean()) #filling the age of test dataset

<p>2 - There are a lot of duplicate values for Ticket (891 passengers and 681 unique values for Tickets (23% of duplicates values)), however Ticker may not be correleted to survival, so we can drop this feature </p>
<p>3 - Does not make sense to use "Name" to predict survival, so we can drop the feature</p>
<p>4 - Does not make sense to use "PassengerID" to predict survival, so we can drop the feature</p>
<p>5 - "Cabin" could be a relevant feature, however there are only 23% of available information (204 of 891), so seems to be better drop this feature</p>

In [39]:
# drop the columns using "drop" method of Pandas
train_df.drop(["Ticket","Name","PassengerId","Cabin"], axis=1, inplace=True) # drop features of train dataset
test_df.drop(["Ticket","Name","PassengerId","Cabin"], axis=1, inplace=True) # drop features of test dataset

<p>6 - Maybe it is a good approach create a new feature named family containing thee sum of "SibSp" and "Parch", this way we will have a feature with the total of family member on board the ship</p>

In [40]:
# Creating the new feature
train_df['Family'] = train_df['SibSp'] + train_df['Parch'] # create the new column in the train dataset
test_df['Family'] = test_df['SibSp'] + test_df['Parch'] # create the new column in the test dataset

<p>Now we can advance with the data analysis to learn more about the features correlation with survival...</p>

In [42]:
# preprocessed dataset train_df
train_df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  Family
0         0       3    male  22.0      1      0   7.2500        S       1
1         1       1  female  38.0      1      0  71.2833        C       1
2         1       3  female  26.0      0      0   7.9250        S       0
3         1       1  female  35.0      1      0  53.1000        S       1
4         0       3    male  35.0      0      0   8.0500        S       0

<h2>Step 3 - Data Analysis</h2>

<h3>Lets try to find correlations between the features and survival</h3>

In [66]:
# Displaying the correlation between the features
train_df.corr()

Survived    Pclass       Age      Fare
Survived  1.000000 -0.338481 -0.069809  0.257307
Pclass   -0.338481  1.000000 -0.331339 -0.549500
Age      -0.069809 -0.331339  1.000000  0.091566
Fare      0.257307 -0.549500  0.091566  1.000000

In [43]:
# analyzing correlation between "Sex" and survival
train_df[["Sex", "Survived"]].groupby(['Sex'], as_index=False).mean().sort_values(by='Survived', ascending=False)

Sex  Survived
0  female  0.742038
1    male  0.188908

<p>It is possible to see that “Sex” has a strong correlation with survival, where 74% of survivors were women</p>

In [44]:
# analyzing correlation between "Pclass" and survival
train_df[['Pclass', 'Survived']].groupby(['Pclass'], as_index=False).mean().sort_values(by='Survived', ascending=False)

Pclass  Survived
0       1  0.629630
1       2  0.472826
2       3  0.242363

In [55]:
# doing statistics analysis using Scipy
print(scipy.stats.pearsonr(train_df['Pclass'], train_df['Survived']))    # Pearson's r method
print(scipy.stats.spearmanr(train_df['Pclass'], train_df['Survived']))   # Spearman's rho  method
print(scipy.stats.kendalltau(train_df['Pclass'], train_df['Survived']))  # Kendall's tau method

(-0.3384810359610153, 2.537047387978631e-25)
SpearmanrResult(correlation=-0.33966793665005257, pvalue=1.6876081409966053e-25)
KendalltauResult(correlation=-0.3235331843940954, pvalue=3.932785644652632e-24)


<p>We can note that 63% of survivors were in the Pclass number 1. The correlation number is about -0.33, what indicates a weak negative correlation between Pclass and survival, however it worths to keep this feature in the analysis</p>

In [46]:
# analyzing correlation between "Family" and survival
train_df[['Family', 'Survived']].groupby(['Family'], as_index=False).mean().sort_values(by='Survived', ascending=False)

Family  Survived
3       3  0.724138
2       2  0.578431
1       1  0.552795
6       6  0.333333
0       0  0.303538
4       4  0.200000
5       5  0.136364
7       7  0.000000
8      10  0.000000

In [52]:
# doing statistics analysis using Scipy
print(scipy.stats.pearsonr(train_df['Family'], train_df['Survived']))    # Pearson's r method
print(scipy.stats.spearmanr(train_df['Family'], train_df['Survived']))   # Spearman's rho  method
print(scipy.stats.kendalltau(train_df['Family'], train_df['Survived']))  # Kendall's tau method

(0.01663898928274522, 0.6198911221336836)
SpearmanrResult(correlation=0.16546291168480362, pvalue=6.822728892835777e-07)
KendalltauResult(correlation=0.15486331354332258, pvalue=7.964587976640051e-07)


<p>Statistically, there is a very weak correlation between "Family" and survival (correlation about 0.1), so lets just drop this column</p>

In [65]:
# drop the columns using "drop" method of Pandas
train_df.drop(["Family","SibSp","Parch"], axis=1, inplace=True) # drop features of train dataset
test_df.drop(["Family","SibSp","Parch"], axis=1, inplace=True) # drop features of test dataset

<h3>Now lets try to visualize the correlations by plotting data</h3>

Survived    Pclass       Age     SibSp     Parch      Fare    Family
Survived  1.000000 -0.338481 -0.069809 -0.035322  0.081629  0.257307  0.016639
Pclass   -0.338481  1.000000 -0.331339  0.083081  0.018443 -0.549500  0.065997
Age      -0.069809 -0.331339  1.000000 -0.232625 -0.179191  0.091566 -0.248512
SibSp    -0.035322  0.083081 -0.232625  1.000000  0.414838  0.159651  0.890712
Parch     0.081629  0.018443 -0.179191  0.414838  1.000000  0.216225  0.783111
Fare      0.257307 -0.549500  0.091566  0.159651  0.216225  1.000000  0.217138
Family    0.016639  0.065997 -0.248512  0.890712  0.783111  0.217138  1.000000

<h2>Step 4 - Model Building</h2>

<h2>Step 5 - Model Evaluation</h2>

<h2>Step 6 - Model Deployment</h2>